Document participant pruning for operators [KVL-421] (#8138)
digital-asset/daml: docs/source/ops/index.rst

.. Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. ops-ref_index:

> Ledger API Server pruning

Reviewer: ```suggestion
DAML Participant pruning
```

Author: Right, I thought we didn't but instead we already use `Participant` throughout the docs.

Author: Done.

=========================

> 

Reviewer: ```suggestion
```suggestion
DAML Participant pruning is currently an :doc:`Early Access Feature in Labs status </support/status-definitions>`

```
```

Author: Done.

> DAML Ledger API servers may support pruning DAML contracts archived and transactions submitted

Author: Done.

> before or at a given ledger offset, for example in order to reduce hot storage usage or to comply

Reviewer: I think it's worth spending an extra sentence or two on the goals of pruning (reduction of hot storage requirements and removal of PII).

Author: Done.

with privacy demands [1]_.

> The DAML Ledger API server pruning functionality is provided by

Reviewer: ```suggestion
The DAML Participant pruning functionality is provided by
```

Author: Done.

the :ref:`Pruning Service <com.daml.ledger.api.v1.PruningService>` Ledger API endpoint,

> requires administrative privileges and it is currently considered EXPERIMENTAL.

Author: Done.

Please refer to the specific DAML Driver information for detailed information about whether it
supports pruning and its impacts.

.. [1] For example, as enabled by provisions about the "right to be forgotten" of legislation such as `EU's GDPR <https://gdpr-info.eu/>`_.

Impacts on DAML applications
----------------------------

> When supported, pruning can be invoked by an operator with administrative privileges at any time on an healthy

Reviewer: ```suggestion
When supported, pruning can be invoked by an operator with administrative privileges at any time on a healthy
```

Author: Done.

> DAML Ledger API server; furthermore, it doesn't require stopping nor suspending normal operation.

Reviewer: ```suggestion
DAML Participant furthermore, it doesn't require stopping nor suspending normal operation.
```

Author: Done.

.. Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. ops-ref_index:

DAML Participant pruning
========================

.. HINT::
   DAML Participant pruning is currently an :doc:`Early Access Feature in Labs status </support/status-definitions>`

The DAML Ledger API exposes an append-only ledger model; on the other hand, DAML Participants must be able to operate continuously for an indefinite amount of time on a limited amount of hot storage.

In addition, privacy demands [1]_ may require removing Personally Identifiable Information (PII) upon request.

> To satisfy these requirements, the :ref:`Pruning Service <com.daml.ledger.api.v1.admin.ParticipantPruningService>` Ledger API endpoint [2]_ allows DAML Participants to support pruning of DAML contracts and transactions that were respectively archived and submitted before or at a given ledger offset.

Reviewer 2: Suggesting to make this more precise to match the specific pruning deletes here, i.e. we prune events and completions (archived contracts are already removed independently form pruning from `participant_contracts`), and we also prune `participant_command_completions`: https://github.com/digital-asset/daml/blob/f0b98a85b9beb06f5fe4619aaba7b0f6c7ae18f0/ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala#L865-L866

The situation with transactions is a bit more subtle when we refer to the `TransactionService.GetTransactionTrees` where we end up potentially returning "partial" transactions with only `CreatedEvent` on behalf of contracts that have not been archived as of the specified pruning offset. But I am not sure how to capture this in a concise way without this being super-confusing.

```suggestion
To satisfy these requirements, the :ref:`Pruning Service <com.daml.ledger.api.v1.admin.ParticipantPruningService>` Ledger API endpoint [2]_ allows DAML Participants to support pruning of events on behalf of DAML contracts and command completions that were respectively archived and completed before or at a given ledger offset.
```

Reviewer: That's not correct though, right? We don't return "partial" transactions. We retain create events in the `participant_events` table (if the respective contract hasn't been archived yet) so that we can serve the ActiveContractsService. But we don't don't deliver these create events as part of the transaction services, because the minimum starting offset of transaction streams must be greater than the last pruned offset.

Reviewer 2: You're right - I forgot that we end up hiding this part. Good. So this speaks to the second paragraph in my comment above - will try to cross that out as the rest is still relevant I think.

Author: @oliverse-da I applied your suggestions about this. Thanks!

Please refer to the specific DAML Driver information for details about its pruning support.

.. [1] For example, as enabled by provisions about the "right to be forgotten" of legislation such as
       `EU's GDPR <https://gdpr-info.eu/>`_.
.. [2] Invoking the Pruning Service requires administrative privileges.

Impacts on DAML applications
----------------------------

When supported, pruning can be invoked by an operator with administrative privileges at any time on a healthy DAML participant; furthermore, it doesn't require stopping nor suspending normal operation.

Still, DAML applications may be affected in the following ways:

> - Pruning is potentially a long-running operation and demanding one in terms of system resources; as such, it may significantly reduce DAML Ledger API throughput and increase latency while it is being performed. It is thus recommended to plan pruning invocations when low system utilization is expected.

Reviewer 2: Good!

```suggestion
- Pruning is potentially a long-running operation and demanding one in terms of system resources; as such, it may significantly reduce DAML Ledger API throughput and increase latency while it is being performed. It is thus strongly recommended to plan pruning invocations when low system utilization is expected.
```

Author: Done.

> - Pruning may degrade the behavior of and/or crash the system if the pruning offset is too recent. In particular, the system might misbehave if command completions are pruned before the command trackers are able to process the completions.

Reviewer 2: Good to call this out specifically as we had discussed with @oggy- . I am not sure if "crashing the system" best captures the worst-case scenario. Isn't the worst that can happen that completions would appear to block indefinitely?

Also is there an impact on command deduplication?

Perhaps this is also a good place to recommend that pruning should be run on "older" history than what is configured for command deduplication and command tracker retention?

Author: > I am not sure if "crashing the system" best captures the worst-case scenario. Isn't the worst that can happen that completions would appear to block indefinitely?

I rephrased to capture this a a slightly more general way.

Author: > Also is there an impact on command deduplication? Perhaps this is also a good place to recommend that pruning should be run on "older" history than what is configured for command deduplication and command tracker retention?

Seems sensible to me. @gerolf-da?

Reviewer: yeah, makes sense. although this value is user supplied, so the participant operator likely should state their pruning window and applications need to adopt their timeouts correspondingly.

Author: Done.

- Pruning may affect the behavior of Ledger API calls that allow to read data from the ledger: see the next sub-section for more information about API impacts.

How the DAML Ledger API is affected
-----------------------------------

> - Active data streams from the DAML Participant may abort and need to be re-established by the DAML application from a later offset than pruned, even if they are already streaming past it.

Reviewer 2: Regarding "even if they are already streaming past it", I believe with the current code, this is only the case in the `CommandCompletions` case, but nevertheless for clarity it is probably good to leave this sentence in here also in case we later find it becomes necessary to abort transaction streams more aggressively too.

Author: Yes, I'd rather stay a bit more vague here and detail these aspects in the API docs, as in this way we won't need to update it if details change (esp. for the worse). They are less relevant for the operator anyway, actually we might also consider removing the API-related sections altogether from the ops guide. WDYT @gerolf-da ?

Reviewer: The operator needs to interact with the API to trigger pruning, so I'm fine with having a more vague representation of the API here, together with a link to the API docs.

Other limitations
-----------------

- If the number of divulged contracts continually grows, the storage used by the participant node may still keep growing, even with regular pruning.
- Pruning may be rejected even if the node is running correctly (for example, to preserve non-repudiation properties); in this case, the application might not be able to archive contracts containing PII or pruning of these contracts may not be possible; thus, actually deleting this PII may also be technically unfeasible.
- Pruning may leave parties, packages, and configuration data on the participant node, even if they are no longer needed for transaction processing, and even if they contain PII [3]_.
- Pruning does not move pruned information to cold storage but simply deletes pruned data; for this reason, it is advisable to back up the Participant Index DB before invoking pruning. See the next sub-section for more Participant Index DB-related advice before and after invoking `prune`.

> - Pruning is not selective but rather effectively truncates the ledger, removing the transactions and archived contracts at and before the pruning offset.

Reviewer 3: - Pruning deletes data from the underlying database and therefore frees up space *within* the database, which can and will be reused during the continued operation of the Index DB. Whether this freed up space is handed back to the OS depends on the database in use. For Postgres for example, the deleted data frees up space in the table storage itself, but does not shrink the size of the files backing the tables of the IndexDB. See the documentation on [`VACUUM` and `VACUUM FULL`](https://www.postgresql.org/docs/12/sql-vacuum.html) for more information..

Reviewer 3: @fabiotudone-da : this comment can be ignored or worked into the comments below. I've written it before seeing that.

Author: I think it is valuable, I'll work it out into the content on the same subject.

Author: Integrated.

> - Pruning is not selective but rather effectively truncates the ledger, removing the transactions and archived contracts at and before the pruning offset.

Reviewer 2: ```suggestion
- Pruning is not selective but rather effectively truncates the ledger, removing the events on behalf of archived contracts and command completions at and before the pruning offset.
```

Author: Done.

.. [3] This might be improved in future versions.

Pruning and the DAML Participant Index DB
-----------------------------------------

Activities to be carried out *before* invoking a pruning operation include:

> - Backing up the Participant Index DB, as pruning will not move information to cold storage but rather it will delete archived contracts and transactions earlier than or at the pruning offset.

Reviewer 2: ```suggestion
- Backing up the Participant Index DB, as pruning will not move information to cold storage but rather it will delete events on behalf of archived contracts and command completions earlier than or at the pruning offset.
```

Author: Done.

Activities to be carried out *after* invoking a pruning operation include:

> - On a PostreSQL Index DB, a manual `VACUUM FULL` command is needed for the DB to give back disk space to the OS.

Reviewer 3: @fabiotudone-da : This is problematic, as it will block all inserts to the Index DB AFAIK.

I'd suggest we do not recommend this, but instead recommend using VACUUM and regular pruning to keep the disk file size bounded.

Author: @gerolf-da I think you suggested `VACUUM FULL`, what do you think? We could also suggest both, e.g. `VACUUM` to avoid downtime and the `FULL` variant when downtime is(/can be) planned to actually give back disk space to the OS.

Reviewer: We probably should put a disclaimer that these are procedures that *can*/*should* be done before/after pruning, but that the user should consult with a "postgresql professional".

Author: Yes, we are definitely entering DB admin territory.

Reviewer 2: Agreed that this is DB admin territory where multiple database server-specific actions may make sense (e.g. another one in case of Postgres: tune auto-vacuum).

Author: I added some more insights after our discussions and I think it is still relatively (enough) light-weight. Let me know WDYT.

Determining a suitable pruning offset
-------------------------------------

The :ref:`Active Contract Service <active-contract-service>` and the :ref:`Transaction Service <transaction-service>` provide offset information for transactions and for Active Contracts snapshots respectively: such offset can be used unchanged with `prune` calls.

Scheduled jobs, applications and/or operator tools can be built on top of the DAML Ledger API to implement pruning automatically, for example at regular intervals, or on-demand, for example according to a user-initiated process.

For instance, pruning at regular intervals could be performed by a cron job that:

1. If a pruning interval has been saved to a well-known location:
   a. Backs up the DAML Participant Index DB.
   b. Performs pruning.
   c. (If using PostgreSQL) Performs a `VACUUM FULL` command on the DAML Participant Index DB.

> 2. Acquires a fresh Active Contract Set and saves the offset.

Reviewer 2: This is for the purpose of knowing the offset to use on one of the subsequent pruning calls, right? I would highlight / track somewhere the use-case for helping the operator determine a suitable offset. E.g. in canton we currently use a [`locate_offset` helper](https://www.canton.io/docs/stable/user-manual/usermanual/console.html#ledger-pruning) which may or may not serve as inspiration for at least an "alpha/preview" level piece of api once we feel sufficiently sure about committing even to a slightly-more-deprecatable piece of admin api perhaps on the `ParticipantPruningService`.

Author: > This is for the purpose of knowing the offset to use on one of the subsequent pruning calls, right?

Exactly right.

> I would highlight / track somewhere the use-case for helping the operator determine a suitable offset.

@gerolf-da I have sent some notes to help us track this, let me know if I can help further.

Reviewer: Typically the offset from the ActiveContractsService is fairly recent. Pruning right away would be too aggressive. I think it's worth pointing this out.

Reviewer: Instead of using the ActiveContractsService to retrieve the offset, it's more efficient to query the ledger end (`TransactionService#GetLedgerEnd`) in regular intervals (e.g. daily at midnight) and store the corresponding offset.

Author: Yep, changed.

Pruning could also be initiated on-demand at the offset of a specific transaction [4]_, for example as provided by a user application based on a search.

> .. [4] Note that not only a specific transaction but also earlier transactions and archived contracts will be pruned.

Reviewer 3: The approach described here to determine an offset is a bit vague, and leaves quite a bit of work to the operator. Can we do better?

I mean with access to the index-db it is a simple query to find an appropriate offset. Can, should, do we expose such information on the Ledger API?

Author: I'd avoid making schema-dependent suggestions, so I'd rather support it at the API level as a feature improvement (and document it then). Not sure if we can improve the present suggestions about determining an offset without that; @meiersi-da @gerolf-da WDYT?

Reviewer: What do you mean by "such information? Transactions have offsets and a record time. We currently don't have a way to scan backwards until you reach an offset that has a certain age.
While one can do that by accessing the index db directly, this won't be very convenient for operational tooling that might not have access to the DB.
If we assume (maybe incorrectly?) that usually users would want a rolling pruning window of x days, then maybe an additional service method that allows for querying the nearest offset to a given timestamp would allow for such operational tooling to be efficiently built with the ledger api. WDYT, @meiersi-da, @fabiotudone-da?

Reviewer: Sorry, didn't see your comment while I was typing mine. I'm in favor of of an additional Ledger API endpoint.

Reviewer: The original participant pruning design also allowed to provide a timestamp in `PruneRequest`. Maybe we should pick this up again and return the pruning offset in the response? It would avoid the extra round trip and seemlessly support the rolling pruning window use case.

Author: I agree that conveniently augmenting the Pruning Ledger API is the ultimate goal but operational tooling could be a good intermediate step and it doesn't need to be tied to (or communicating with) the DB directly, for example it could leverage an internal admin API. Or we could expose experimental APIs although I am a bit wary of the de-facto dependencies that we'd allow in that case.

Another option, after discussing with @oliverse-da (thanks!), would be to have the operator determine the offset by adding a specific amount to either the ledger's start, if not pruned, or to the most recent available offset returned as part of the error message when querying about the ledger's start (and, to avoid exposing the message format, we could make the offset within available through tooling). That amount would be sized based on the deployment's events rate.

It looks like all improved suggestions require some additional development though, so I'd leave this part of the doc as it is for now.

I'd also suggest to keep involving @oliverse-da and team (subject to their availability, of course) for the upcoming improvements and developments as they have done a great deal of thinking, design, experiments, measurements and have overall a lot of experience with pruning towards significant production-ready deployments.

Reviewer 2: For multi-domain ledgers time will not necessarily map unambiguously to offsets as time does not grow monotonically wrt offsets. (cc @oggy- ) - see my note [above](https://github.com/digital-asset/daml/pull/8138/files#r535989613) for how we currently fill the gap until we have a ledger-api-way.

Author: @oliverse-da I replaced "earlier" occurrences with "at previous offsets" or similar. Will this do?

Reviewer 2: > @oliverse-da I replaced "earlier" occurrences with "at previous offsets" or similar. Will this do?

Sorry Fabio, my reply in the previous suggestion was in response to Gerolf's reference to the old "PruneByTime" flavor we used to support in canton, but had gotten rid of because of time-semantic reasons. I did not intend for you to rephrase "earlier" with "at previous offsets" although that seems good.

.. ops-ref_index:

DAML Participant pruning
========================

.. HINT::
   DAML Participant pruning is currently an :doc:`Early Access Feature in Labs status </support/status-definitions>`

The DAML Ledger API exposes an append-only ledger model; on the other hand, DAML Participants must be able to operate continuously for an indefinite amount of time on a limited amount of hot storage.

> In addition, privacy demands [1]_ may require to remove removing Personally Identifiable Information (PII) upon request.

Reviewer: ```suggestion
In addition, privacy demands [1]_ may require removing Personally Identifiable Information (PII) upon request.
```

Author: Done.

- Pruning may be rejected even if the node is running correctly (for example, to preserve non-repudiation properties); in this case, the application might not be able to archive contracts containing PII or pruning of these contracts may not be possible; thus, actually deleting this PII may also be technically unfeasible.
- Pruning may leave parties, packages, and configuration data on the participant node, even if they are no longer needed for transaction processing, and even if they contain PII [3]_.
- Pruning does not move pruned information to cold storage but simply deletes pruned data; for this reason, it is advisable to back up the Participant Index DB before invoking pruning. See the next sub-section for more Participant Index DB-related advice before and after invoking `prune`.
- Pruning is not selective but rather effectively truncates the ledger, removing events on behalf of archived contracts and command completions at the pruning offset and all previous offsets.

.. [3] This might be improved in future versions.

How Pruning affects Index DB administration
-------------------------------------------

> Pruning deletes data from the underlying database and therefore frees up space within it, which can and will be reused during the continued operation of the Index DB. Whether this freed up space is handed back to the OS depends on the database in use. For example, in PostgreSQL the deleted data frees up space in the table storage itself, but does not shrink the size of the files backing the tables of the IndexDB. Please refer to the PostgreSQL documentation on `VACUUM` and `VACUUM FULL` for more information.

Author: Done.

Activities to be carried out *before* invoking a pruning operation should thus include backing up the Participant Index DB, as pruning will not move information to cold storage but rather it will delete events on behalf of archived contracts and command completions before or at the pruning offset.

In addition, activities to be carried out *after* invoking a pruning operation might include:

> - On a PostreSQL Index DB, especially if auto-vacuum tuning has not been performed, issuing `VACUUM` commands at appropriate times may improve performance and storage usage by letting the DB reuse freed space. Note that `VACUUM FULL` commands are still needed for the OS to reclaim disk space previously used by the DB.

Author: Done.

Backing up and vacuuming, in addition to pruning itself, are also long-running and resource-hungry operations that might negatively affect the performance of regular workloads and even the availability of the system: this is true in particular for `VACUUM FULL` in PostgreSQL and equivalent commands in other DBMSs. These operations should thus be planned and taken carefully into account when sizing system resources. They should also be scheduled sensibly in relation to the desired sustained performance levels of regular workloads and to the hot storage usage goals.

> Professional advice on DB administration is strongly recommended that would take into account the DB specifics as well as all of the above aspects.

Author: Done.
